File: ios/SneakBit/Analytics/RuntimeEvents.swift
import Combine
import Foundation

protocol RuntimeEventsBroker {
    func send(_ event: RuntimeEvent)
    func events() -> AnyPublisher<RuntimeEvent, Never>
}

enum RuntimeEvent {
    case loading
    case launched
    case willEnterForeground
    case didEnterBackground
    case gameOver
    case newGame
    case worldTransition(source: UInt32, destination: UInt32)
}

class RuntimeEventsBrokerImpl: RuntimeEventsBroker, Loggable {
    private let latestEvent = CurrentValueSubject<RuntimeEvent, Never>(.loading)
    
    func send(_ event: RuntimeEvent) {
        latestEvent.send(event)
        log(event)
    }
    
    func events() -> AnyPublisher<RuntimeEvent, Never> {
        latestEvent.eraseToAnyPublisher()
    }
    
    private func log(_ event: RuntimeEvent) {
        log(event.description)
    }
}

extension RuntimeEvent: CustomStringConvertible {
    var description: String {
        switch self {
        case .loading: "Loading..."
        case .launched: "Launched!"
        case .willEnterForeground: "Entering foreground"
        case .didEnterBackground: "Entered background"
        case .gameOver: "Game Over"
        case .newGame: "Started new game"
        case .worldTransition(let source, let destination): "World changed from \(source) to \(destination)"
        }
    }
}

File: android/app/src/main/java/it/curzel/bitscape/analytics/RuntimeEvents.kt
package it.curzel.bitscape.analytics

import kotlinx.coroutines.flow.*
import android.util.Log

class RuntimeEventsBroker {
    private val latestEvent = MutableStateFlow<RuntimeEvent>(RuntimeEvent.Loading)

    companion object {
        private const val TAG = "RuntimeEventsBroker"
    }

    fun send(event: RuntimeEvent) {
        latestEvent.value = event
        log(event.description)
    }

    fun events(): Flow<RuntimeEvent> = latestEvent.asStateFlow()

    private fun log(message: String) {
        Log.d(TAG, message)
    }
}

sealed class RuntimeEvent {
    object Loading : RuntimeEvent()
    object Launched : RuntimeEvent()
    object WillEnterForeground : RuntimeEvent()
    object DidEnterBackground : RuntimeEvent()
    object GameOver : RuntimeEvent()
    object NewGame : RuntimeEvent()
    data class WorldTransition(val source: UInt, val destination: UInt) : RuntimeEvent()

    val description: String
        get() = when (this) {
            is Loading -> "Loading..."
            is Launched -> "Launched!"
            is WillEnterForeground -> "Entering foreground"
            is DidEnterBackground -> "Entered background"
            is GameOver -> "Game Over"
            is NewGame -> "Started new game"
            is WorldTransition -> "World changed from ${source} to ${destination}"
        }
}
